crypto/init.c
=======================================================
lhs: 100644 | 1fa5e894208e958b755095ebb97bf03bb7a34858
rhs: 100644 | d50d7f19e8eb312746a24e72f3771d2c18a2d491
---@@ -474,12 +474,17 @@ void OPENSSL_cleanup(void)
                     "RAND_cleanup()\n");
 
 #endif
-    CRYPTO_cleanup_all_ex_data();
-    EVP_cleanup();
-    CONF_modules_free();
+/*
+ * Note that cleanup order is important.
+ * For example, ENGINEs use CRYPTO_EX_DATA and therefore, must be cleaned up
+ * before the ex data handlers are wiped in CRYPTO_cleanup_all_ex_data().
+ */
 #ifndef OPENSSL_NO_ENGINE
     ENGINE_cleanup();
 #endif
+    CRYPTO_cleanup_all_ex_data();
+    EVP_cleanup();
+    CONF_modules_free();
     RAND_cleanup();
     base_inited = 0;
 }
@@ -628,5 +633,3 @@ int OPENSSL_atexit(void (*handler)(void))
 
     return 1;
 }
-
-

---